Ambiguous glob import of `main` errors even in library crate

I compiled the following code with `rustc --crate-type=lib repro.rs`

```rust
mod a {
    pub fn main() {}
}
mod b {
    pub fn main() {}
}

use a::*;
use b::*;
```

I expected the code to compile, but instead I got the following error:

```
error[E0659]: `main` is ambiguous
  |
  = note: ambiguous because of multiple glob imports of a name in the same module
note: `main` could refer to the function imported here
 --> repro.rs:8:5
  |
8 | use a::*;
  |     ^^^^
  = help: consider adding an explicit import of `main` to disambiguate
note: `main` could also refer to the function imported here
 --> repro.rs:9:5
  |
9 | use b::*;
  |     ^^^^
  = help: consider adding an explicit import of `main` to disambiguate

warning: unused import: `b::*`
 --> repro.rs:9:5
  |
9 | use b::*;
  |     ^^^^
  |
  = note: `#[warn(unused_imports)]` (part of `#[warn(unused)]`) on by default

error: aborting due to 1 previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0659`.
```

In a library crate, rustc shouldn't be caring whether a function is named `main` or not.

Renaming both functions from `main` to `foo` makes the code compile without errors.

Possibly relevant code: https://github.com/rust-lang/rust/blob/88bd39beb342e2ae7ded1eb7d58697416686d679/compiler/rustc_resolve/src/lib.rs#L1875

### Meta


`rustc --version --verbose`:
```
rustc 1.93.0-nightly (80d8f292d 2025-11-25)
binary: rustc
commit-hash: 80d8f292d82d735f83417221dd63b0dd2bbb8dd2
commit-date: 2025-11-25
host: aarch64-apple-darwin
release: 1.93.0-nightly
LLVM version: 21.1.5
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -73,6 +73,7 @@ use rustc_middle::ty::{
     ResolverGlobalCtxt, TyCtxt, TyCtxtFeed, Visibility,
 };
 use rustc_query_system::ich::StableHashingContext;
+use rustc_session::config::CrateType;
 use rustc_session::lint::builtin::PRIVATE_MACRO_USE;
 use rustc_span::hygiene::{ExpnId, LocalExpnId, MacroKind, SyntaxContext, Transparency};
 use rustc_span::{DUMMY_SP, Ident, Macros20NormalizedIdent, Span, Symbol, kw, sym};
@@ -2430,6 +2431,12 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
     }
 
     fn resolve_main(&mut self) {
+        let any_exe = self.tcx.crate_types().contains(&CrateType::Executable);
+        // Don't try to resolve main unless it's an executable
+        if !any_exe {
+            return;
+        }
+
         let module = self.graph_root;
         let ident = Ident::with_dummy_span(sym::main);
         let parent_scope = &ParentScope::module(module, self.arenas);

--- a/tests/ui/entry-point/imported_main_conflict_lib.rs
+++ b/tests/ui/entry-point/imported_main_conflict_lib.rs
@@ -1,0 +1,10 @@
+// Tests that ambiguously glob importing main doesn't fail to compile in non-executable crates
+// Regression test for #149412
+//@ check-pass
+#![crate_type = "lib"]
+
+mod m1 { pub(crate) fn main() {} }
+mod m2 { pub(crate) fn main() {} }
+
+use m1::*;
+use m2::*;

